### PR TITLE
support sampleRate as Dynamic Configuration (#4968)

### PR DESCRIPTION
**Pull request by killGC**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Comment by wu-sheng** [comment]
Please fix the CI first.

**Comment by wu-sheng** [comment]
There is code format style file in the root folder, codeStyle.xml. And you could run these locally.
1. `./mvnw clean deploy`
1. `./mvnw apache-rat:check`

**Comment by killGC** [comment]
> There is code format style file in the root folder, codeStyle.xml. And you could run these locally.
> 
> 1. `./mvnw clean deploy`
> 2. `./mvnw apache-rat:check`

when execute `./mvnw clean deploy`,appear “Unauthorized” error

**Comment by killGC** [comment]
> > There is code format style file in the root folder, codeStyle.xml. And you could run these locally.
> > 
> > 1. `./mvnw clean deploy`
> > 2. `./mvnw apache-rat:check`
> 
> when execute `./mvnw clean deploy`,appear “Unauthorized” error

截图5

**Comment by killGC** [comment]
i execute mvnw package,appear errors

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project apm-network: Compilation failure: Compilation failure:
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/ProfileTaskCommand.java:[21,51] 程序包org.apache.skywalking.apm.network.common.v3
不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/ProfileTaskCommand.java:[22,51] 程序包org.apache.skywalking.apm.network.common.v3
不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/BaseCommand.java:[21,51] 程序包org.apache.skywalking.apm.network.common.v3不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/BaseCommand.java:[22,51] 程序包org.apache.skywalking.apm.network.common.v3不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/Serializable.java:[21,51] 程序包org.apache.skywalking.apm.network.common.v3不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/Deserializable.java:[21,51] 程序包org.apache.skywalking.apm.network.common.v3不存
在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/ProfileTaskCommand.java:[54,43] 找不到符号
[ERROR]   符号:   类 Command
[ERROR]   位置: 类 org.apache.skywalking.apm.network.trace.component.command.ProfileTaskCommand
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/ProfileTaskCommand.java:[92,19] 程序包Command不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/BaseCommand.java:[28,26] 程序包Command不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/BaseCommand.java:[43,12] 程序包Command不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/Serializable.java:[24,12] 程序包Command不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/Deserializable.java:[24,19] 找不到符号
[ERROR]   符号:   类 Command
[ERROR]   位置: 接口 org.apache.skywalking.apm.network.trace.component.command.Deserializable<T>
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/TraceIgnoreCommand.java:[21,51] 程序包org.apache.skywalking.apm.network.common.v3
不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/TraceIgnoreCommand.java:[22,51] 程序包org.apache.skywalking.apm.network.common.v3
不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/TraceIgnoreCommand.java:[34,19] 程序包Command不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/CommandDeserializer.java:[20,51] 程序包org.apache.skywalking.apm.network.common.v3
不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/CommandDeserializer.java:[24,49] 找不到符号
[ERROR]   符号:   类 Command
[ERROR]   位置: 类 org.apache.skywalking.apm.network.trace.component.command.CommandDeserializer
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/UnsupportedCommandException.java:[20,51] 程序包org.apache.skywalking.apm.network.c
ommon.v3不存在
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/UnsupportedCommandException.java:[23,19] 找不到符号
[ERROR]   符号:   类 Command
[ERROR]   位置: 类 org.apache.skywalking.apm.network.trace.component.command.UnsupportedCommandException
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/UnsupportedCommandException.java:[25,46] 找不到符号
[ERROR]   符号:   类 Command
[ERROR]   位置: 类 org.apache.skywalking.apm.network.trace.component.command.UnsupportedCommandException
[ERROR] /E:/workspace/killGC-skywalking/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/UnsupportedCommandException.java:[29,12] 找不到符号
[ERROR]   符号:   类 Command
[ERROR]   位置: 类 org.apache.skywalking.apm.network.trace.component.command.UnsupportedCommandException
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :apm-network

**Comment by arugal** [comment]
> > There is code format style file in the root folder, codeStyle.xml. And you could run these locally.
> > 
> > 1. `./mvnw clean deploy`
> > 2. `./mvnw apache-rat:check`
> 
> when execute `./mvnw clean deploy`,appear “Unauthorized” error

Please try to `./mvnw -DskipTests clean package`

**Comment by killGC** [comment]
> > > There is code format style file in the root folder, codeStyle.xml. And you could run these locally.
> > > 
> > > 1. `./mvnw clean deploy`
> > > 2. `./mvnw apache-rat:check`
> > 
> > 
> > when execute `./mvnw clean deploy`,appear “Unauthorized” error
> 
> Please try to `./mvnw -DskipTests clean package`

截图6

**Comment by killGC** [comment]
> > > > There is code format style file in the root folder, codeStyle.xml. And you could run these locally.
> > > > 
> > > > 1. `./mvnw clean deploy`
> > > > 2. `./mvnw apache-rat:check`
> > > 
> > > 
> > > when execute `./mvnw clean deploy`,appear “Unauthorized” error
> > 
> > 
> > Please try to `./mvnw -DskipTests clean package`
> 
ERROR org.apache.skywalking.apm.network.common.v3不存在
“org.apache.skywalking.apm.network.common.v3” is in “apm-protocol” module？
截图7

**Comment by arugal** [comment]
@killGC You need to initialize the Git submodule, [Please follow this documentation](https://github.com/apache/skywalking/blob/master/docs/en/guides/How-to-build.md#build-from-github).

**Comment by wu-sheng** [comment]
Still have code style issues.

**Comment by killGC** [comment]
> Still have code style issues.

local compile and package speed is so slow,because  need to download so much dependency jar file from maven depository

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            sampleRate.set(Integer.parseInt(config));
+        } catch (NumberFormatException ex) {
+            log.error("Cannot load sampleRate from: {}", config, ex);
+        }
+    }
+
+    @Override
+    public void notify(ConfigChangeEvent value) {
+        if (EventType.DELETE.equals(value.getEventType())) {
+            activeSetting("");
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:56]: What is the expected sampling rate if the delete happens?

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceModuleProvider.java [hunk 2]
```diff
@@ -75,8 +77,11 @@ public void prepare() throws ServiceNotProvidedException {
 
         uninstrumentedGatewaysConfig = new UninstrumentedGatewaysConfig(this);
 
+        traceSampleRateWatcher = new TraceSampleRateWatcher(String.valueOf(moduleConfig.getSampleRate()), this);
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceModuleProvider.java:80]: I think should not change to String here, please move this into the `TraceSampleRateWatcher`.

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.apache.skywalking.oap.server.configuration.api.ConfigChangeWatcher;
+import org.apache.skywalking.oap.server.core.Const;
+import org.apache.skywalking.oap.server.receiver.trace.module.TraceModule;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+@Slf4j
+public class TraceSampleRateWatcher extends ConfigChangeWatcher {
+    private  AtomicReference<String> settingsString;
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:30]: I think you don't need `settingsString`

Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:30]: i think so too

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java (hunk 1 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:56]: change back to property sampleRate of class TraceServiceModuleConfig

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceModuleProvider.java (hunk 2 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceModuleProvider.java:80]: ok

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java (hunk 3 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:30]: every time execute "./mvnw -DskipTests clean package" command so long,What's the best way to improve or whether or not have some way only execute changed module

Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:30]: image

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+package org.apache.skywalking.oap.server.receiver.trace.provider;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.skywalking.oap.server.configuration.api.ConfigChangeWatcher;
+import org.apache.skywalking.oap.server.receiver.trace.module.TraceModule;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+@Slf4j
+public class TraceSampleRateWatcher extends ConfigChangeWatcher {
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:28]: Please set up a UT for this. Test this watcher could be driven correctly.

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            activeSetting(value.getNewValue());
+        }
+    }
+
+    @Override
+    public String value() {
+        return String.valueOf(sampleRate.get());
+    }
+
+    public int getDefaultValue() {
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:62]: Why public?

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        return ((TraceModuleProvider) this.getProvider()).getModuleConfig().getSampleRate();
+    }
+
+    public int getSampleRate() {
+        return sampleRate.get() == null ? getDefaultValue() : sampleRate.get();
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:67]: I think you should use `sampleRate.get()` only. Why `sampleRate.get() == null`? In which case?

Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:67]: great，thank you for help

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java (hunk 5 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:62]: i see，sorry for  the problem

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java (hunk 4 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:28]: ok

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.skywalking.oap.server.configuration.api.ConfigChangeWatcher;
+import org.apache.skywalking.oap.server.receiver.trace.module.TraceModule;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+@Slf4j
+public class TraceSampleRateWatcher extends ConfigChangeWatcher {
+    private  AtomicReference<Integer> sampleRate;
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:29]:
```suggestion
    private AtomicReference<Integer> sampleRate;
```

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java (hunk 4 above)
Comment (wu-sheng) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:28]: @killGC Could you finish the UT codes ASAP, I have to wait that before merging.

**Comment by wu-sheng** [comment]
@songzhendong Could you test this locally? As we don't have e2e for this today, and you are playing the config center these days.

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java (hunk 4 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcher.java:28]: i'm not familiar with UT at here,i need spend some time to study,

**Comment by songzhendong** [comment]
My computer compile error, please check the code
image
image

**Comment by killGC** [comment]
> My computer compile error, please check the code
> image
> image
> <file name="/home/songzhendong/killGC/skywalking/oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/test/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcherTest.java">
> <error line="64" column="73" severity="error" message="',' is not followed by whitespace." source="com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck"/>

ok,thank you

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/test/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcherTest.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+public class TraceSampleRateWatcherTest {
+    private TraceModuleProvider traceModuleProvider;
+
+    @Before
+    public void init() {
+        traceModuleProvider = new TraceModuleProvider();
+    }
+
+    @Test
+    public void TestInit() {
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/test/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcherTest.java:46]:
```suggestion
    public void testInit() {
```

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/test/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcherTest.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+        while (watcher.getSampleRate() == 10000) {
+            Thread.sleep(2000);
+        }
+        assertThat(watcher.getSampleRate(), is(9000));
+        assertThat(traceModuleProvider.getModuleConfig().getSampleRate(), is(10000));
+    }
+
+    @Test
+    public void TestNotify() {
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/test/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcherTest.java:68]:
```suggestion
    public void testNotify() {
```

Comment (wu-sheng) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/test/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceSampleRateWatcherTest.java:68]: FYI @kezhenxu94 Could you check why the method name format didn't check this?

**Comment by songzhendong** [comment]
SampleRate as dynamic configuration detection passed ,it works.

**Comment by songzhendong** [comment]
Dynamic configuration sampling rate test passed ,it works.

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java [hunk 10]
```diff
@@ -18,18 +18,20 @@
...
 
+import org.apache.skywalking.oap.server.receiver.trace.provider.TraceServiceModuleConfig;
+
 /**
  * The sampler makes the sampling mechanism works at backend side. Sample result: [0,sampleRate) sampled, (sampleRate,~)
  * ignored
  */
 public class TraceSegmentSampler {
-    private int sampleRate = 10000;
+    private TraceServiceModuleConfig config;
```
[the remark this answers is not in the document]
Comment (wu-sheng) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java:28]: @killGC Could you finish this? then it is ready to merge.

[the remark this answers is not in the document]
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java]: yes,i have finished

[the remark this answers is not in the document]
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java]: public class TraceSegmentSampler {
    private **volatile** TraceServiceModuleConfig config;

i modify like this,or
public class TraceSampleRateWatcher extends ConfigChangeWatcher {
    private **volatile** AtomicReference<Integer> sampleRate;

**Comment by killGC** [comment]
> Dynamic configuration sampling rate test passed ,it works.

@songzhendong,thanks

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java [hunk 11]
```diff
@@ -25,7 +25,7 @@
  * ignored
  */
 public class TraceSegmentSampler {
-    private TraceServiceModuleConfig config;
+    private volatile TraceServiceModuleConfig config;
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java:28]: This change is not right. Please read my review more carefully.

Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java:28]: Dynamic configuration will not modify TraceServiceModuleConfig#sampleRate，but modify TraceSampleRateWatcher#sampleRate

Comment (wu-sheng) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java:28]: No one will change this reference, @killGC

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceServiceModuleConfig.java [hunk 12]
```diff
@@ -51,6 +51,9 @@
     @Setter
     @Getter
     private UninstrumentedGatewaysConfig uninstrumentedGatewaysConfig;
+    @Setter
+    @Getter
+    private TraceSampleRateWatcher traceSampleRateWatcher;
```
Comment (wu-sheng) [inline oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceServiceModuleConfig.java:56]: Why config has reference to watcher? This is strange. Watcher should be a part of `TraceSegmentSampler`

Comment (wu-sheng) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceServiceModuleConfig.java:56]: Config should only be changed through the initialization stage. Nothing more.

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java (hunk 11 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java:28]: okay，but this reference #traceSampleRateWatcher#sampleRate will change，to get newest sampleRate's value by this reference，no need for this reference add volatile？

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceServiceModuleConfig.java (hunk 12 above)
Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceServiceModuleConfig.java:56]: i understand this mean this config have which watchers will modify this config‘s properties？for example,same with uninstrumentedGatewaysConfig be added before

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java (hunk 11 above)
Comment (wu-sheng) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java:28]: You are always using the same `config#getTraceSampleRateWatcher ` reference, right? As the `config` is never changed, you don't need this. `volatile` could cause more CPU, please use it if necessary.

Comment (killGC) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/TraceSegmentSampler.java]: okay,i see

File: oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceServiceModuleConfig.java (hunk 12 above)
Comment (wu-sheng) [reply oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/TraceServiceModuleConfig.java:56]: OK, ignore this. Continue on others.

**Comment by killGC** [comment]
> LGTM. Thanks for your patient. Welcome to join us as the 240th or 241st contributor.

it ‘s My pleasure，Thanks for patient too

**Pull request merged**